### ISSUE TITLE
TypeError: Cannot read property 'shift' of null

### ISSUE THREAD

**Issue by AgentME** [rpominov/kefir#134]
In our application, we catch all uncaught exceptions and have the client log them to our server automatically. We've been seeing this many times a day from many users without much else in common:

Error logged: TypeError: Cannot read property 'shift' of null 
    at mixin.\_init.\_$shiftBuff ([src/one-source/delay.js:8](https://github.com/rpominov/kefir/blob/0339735b451a0ed0f93f59e0563f04dc4f8a390c/src/one-source/delay.js#L8))

I've seen it happen a few times randomly, but I haven't been able to dependably reproduce the issue until today. My reproduction steps involved triggering a few certain bugs in the application while having Chrome's debugger set to pause on exceptions, and then this exception happens afterward. (I don't assume this is how users are triggering it.) To get more information, I modified delay.js's code to have a few lines added:

```javascript
const mixin = {
  //...
  _free() {
    this.__free_stack = new Error().stack; // +
    this._buff = null;
    this._$shiftBuff = null;
  },
  //...
  _handleEnd() {
    if (this._activating) {
      this.__immediate_end = true; // +
      this._emitEnd();
    } else {
      setTimeout(() => {
        this.__timeout_end = true; // +
        this._emitEnd();
      }, this._wait);
    }
  }

};
```

At the moment of the exception, `this.__timeout_end` was true, and `this.__free_stack` contained:

Error 
    at AnonymousObservable._free ([src/one-source/delay.js:12](https://github.com/rpominov/kefir/blob/0339735b451a0ed0f93f59e0563f04dc4f8a390c/src/one-source/delay.js#L12)) 
    at AnonymousObservable._clear ([src/patterns/one-source.js:52](https://github.com/rpominov/kefir/blob/0339735b451a0ed0f93f59e0563f04dc4f8a390c/src/patterns/one-source.js#L52)) 
    at AnonymousObservable._emitEnd ([src/property.js:41](https://github.com/rpominov/kefir/blob/0339735b451a0ed0f93f59e0563f04dc4f8a390c/src/property.js#L41)) 
    at [src/one-source/delay.js:29](https://github.com/rpominov/kefir/blob/0339735b451a0ed0f93f59e0563f04dc4f8a390c/src/one-source/delay.js#L29)

It looks like the function assigned to `this._$shiftBuff` is being called after `this._emitEnd`. It looks like the code is relying on setTimeout callbacks with the same wait value being called in the order they're set, but this is not always the case: http://stackoverflow.com/a/5998338/1289657. If you view the example jsfiddle in Chrome, then you can easily trigger an error by hitting the pause and play buttons on the debugger a few times if one doesn't happen on its own.

Here's an example test case which replaces setTimeout with a function more prone to executing events out of order and fails about 20% of the time with the "Cannot read property 'shift' of null" error.

```javascript
var Kefir = require('kefir');

(function() {
  var originalST = global.setTimeout;
  global.setTimeout = function() {
    if (Math.random() > 0.5) {
      return originalST.apply(this, arguments);
    } else {
      var _this = this, _arguments = arguments;
      originalST(function() {
        return originalST.apply(_this, _arguments);
      }, 1);
    }
  };
})();

Kefir.sequentially(0, [1,2,3]).delay(0).log();
```

**Comment by rpominov** [comment]
Hey, Chris!
Thank you for so detailed report! Hopefully I'll be able to look into it at this weekend.

**Comment by pixelpicosean** [comment]
I also had this problem before, but in my case [reactive plugin for a game engine](https://github.com/LesserPanda-Engine/LP-Reactive) a customed Timer is used to replace the setTimeout. When I use `delay` somewhere, it also randomly throws shift exceptions.

So I modified the `delay` to call the end method manually, you can [see it here](https://github.com/LesserPanda-Engine/LP-Reactive/blob/master/reactive.js#L2498-L2520).

Hope that will help :D

**Comment by AgentME** [comment]
If I'm reading the [standard](https://html.spec.whatwg.org/multipage/webappapis.html#timer-initialisation-steps) correctly, specifically step 12, then Kefir's setTimeout handling should be perfectly legal. It's just that Chrome and Safari have bugs that cause them to be executed out of order sometimes. See https://code.google.com/p/chromium/issues/detail?id=172587 and https://bugs.webkit.org/show_bug.cgi?id=111143. Neither example fails for me as-is in current versions unless I ever use the debugger's pause button. I assume there's other things that can trigger it too because I'd be surprised if all of the error logs I've seen regularly happening were all users using the debugger.

**Comment by AgentME** [comment]
I found a way to fix it that isn't a massive hack and put in a PR. Even if there weren't any setTimeout-related bugs, this code would still be a sensible implementation of delay.

**Comment by rpominov** [comment]
Fixed in `2.8.1`. Thanks a lot for the find, @AgentME!
